Userland/Services/WindowServer/Screen.cpp

```diff
@@ -19,66 +19,133 @@
 
 namespace WindowServer {
 
-static Screen* s_the;
+NonnullOwnPtrVector<Screen, 8> Screen::s_screens;
+Screen* Screen::s_main_screen { nullptr };
+Gfx::IntRect Screen::s_bounding_screens_rect {};
 
-Screen& Screen::the()
+ScreenInput& ScreenInput::the()
 {
-    VERIFY(s_the);
-    return *s_the;
+    static ScreenInput s_the;
+    return s_the;
 }
 
-Screen::Screen(unsigned desired_width, unsigned desired_height, int scale_factor)
+Screen& ScreenInput::cursor_location_screen()
 {
-    VERIFY(!s_the);
-    s_the = this;
-    m_framebuffer_fd = open("/dev/fb0", O_RDWR | O_CLOEXEC);
+    auto* screen = Screen::find_by_location(m_cursor_location);
+    VERIFY(screen);
+    return *screen;
+}
+
+const Screen& ScreenInput::cursor_location_screen() const
+{
+    auto* screen = Screen::find_by_location(m_cursor_location);
+    VERIFY(screen);
+    return *screen;
+}
+
+Screen::Screen(const String& device, const Gfx::IntRect& virtual_rect, int scale_factor)
+    : m_virtual_rect(virtual_rect)
+    , m_scale_factor(scale_factor)
+{
+    m_framebuffer_fd = open(device.characters(), O_RDWR | O_CLOEXEC);
     if (m_framebuffer_fd < 0) {
-        perror("failed to open /dev/fb0");
+        perror(String::formatted("failed to open {}", device).characters());
         VERIFY_NOT_REACHED();
     }
 
     if (fb_set_buffer(m_framebuffer_fd, 0) == 0) {
         m_can_set_buffer = true;
     }
 
-    set_resolution(desired_width, desired_height, scale_factor);
-    m_physical_cursor_location = physical_rect().center();
+    // If the cursor is not in a valid screen (yet), force it into one
+    dbgln("Screen() current physical cursor location: {} rect: {}", ScreenInput::the().cursor_location(), rect());
+    if (!find_by_location(ScreenInput::the().cursor_location()))
+        ScreenInput::the().set_cursor_location(rect().center());
 }
 
 Screen::~Screen()
 {
     close(m_framebuffer_fd);
 }
 
-bool Screen::set_resolution(int width, int height, int new_scale_factor)
+void Screen::init()
+{
+    do_set_resolution(true, m_virtual_rect.width(), m_virtual_rect.height(), m_scale_factor);
+}
+
+Screen& Screen::closest_to_rect(const Gfx::IntRect& rect)
 {
+    Screen* best_screen = nullptr;
+    int best_area = 0;
+    for (auto& screen : s_screens) {
+        auto r = screen.rect().intersected(rect);
+        int area = r.width() * r.height();
+        if (!best_screen || area > best_area) {
+            best_screen = &screen;
+            best_area = area;
+        }
+    }
+    if (!best_screen) {
+        // TODO: try to find the best screen in close proximity
+        best_screen = &Screen::main();
+    }
+    return *best_screen;
+}
+
+Screen& Screen::closest_to_location(const Gfx::IntPoint& point)
+{
+    for (auto& screen : s_screens) {
+        if (screen.rect().contains(point))
+            return screen;
+    }
+    // TODO: guess based on how close the point is to the next screen rectangle
+    return main();
```

Comment:
Okay, so that's _technically_ correct but it was really confusing for me at first glance ^^' How about caling `Screen::main()` explicitly ? (I don't even know if it compiles so feel free to ignore this if it doesn't)